### ISSUE TITLE
Make node-oracledb installable in Windows

### ISSUE THREAD

**Issue by bchr02** [oracle/node-oracledb#7]
I know that you have already mentioned that this is coming in the near future, but since I am very eager for it and so that I can get alerted when it's actually completed, I am creating this issue.

Perhaps an ETA can be provided? Thank you.

**Comment by cjbj** [comment]
No ETA. Do you have experience with node-gyp on Windows?

**Comment by b2dac** [comment]
No but I am willing to read the documentation if you need help. What can I do?

I tried installing it but got an error message pointing to gyp file complaining about oci_lib_dir not being defined. So I opened the binding.gyp file and noticed there was no condition for "win". I created one and hard coded the paths and then I got a little further in the install process but it ended up crashing again at another point. Regarding the gyp file, I think what we need to do is point it to env variables for both the include and lib directories. Is this what you need help with? If so, I'm sure I can do that.

By the way, I found this site which looks like a good reference guide: https://code.google.com/p/gyp/wiki/InputFormatReference

**Comment by rinie** [comment]
Ok, got examples\connect.js and examples.select1.js working under 32bits vc12 with windows 8.1 and instantclient_12_1.
Do'nt know if all modifications that I did were necessary, but:
1. Source code modification:
dpiException.h: msvc has a macro errno(), so I added 
#undef errno

2. Path node-oracledb for module oracledb gives module load error:
Renamed node-oracledb directory to oracledb
Installed instant client in
C:\Temp\node\node_modules\oracledb\instantclient_12_1
Add this to path:
set path=C:\Temp\node\node_modules\oracledb\instantclient_12_1;%PATH%

3. Binding.gyp settings taken from node-oracle:
{
  "targets": [
  {
    "target_name" : "oracledb",
    "sources" : [
             "src/njs/src/njsOracle.cpp",
             "src/njs/src/njsPool.cpp",
             "src/njs/src/njsConnection.cpp",
             "src/njs/src/njsMessages.cpp",             
             "src/dpi/src/dpiEnv.cpp",
             "src/dpi/src/dpiEnvImpl.cpp",
             "src/dpi/src/dpiException.cpp",
             "src/dpi/src/dpiExceptionImpl.cpp",
             "src/dpi/src/dpiConnImpl.cpp",
             "src/dpi/src/dpiDateTimeArrayImpl.cpp",
             "src/dpi/src/dpiPoolImpl.cpp",
             "src/dpi/src/dpiStmtImpl.cpp",
             "src/dpi/src/dpiUtils.cpp"
    ],
    "conditions" : [
    [
      'OS=="linux"', {
        "variables" : {
          "oci_inc_dir%" : "<!(if [ -z $OCI_INC_DIR ]; then echo \"/usr/include/oracle/12.1/client64/\"; else echo $OCI_INC_DIR; fi)",
          "oci_lib_dir%" : "<!(if [ -z $OCI_LIB_DIR ]; then echo \"/usr/lib/oracle/12.1/client64/lib/\"; else echo $OCI_LIB_DIR; fi)",
        }
      }
    ],
    [
      'OS=="mac"', {
        "xcode_settings": {
	  "GCC_ENABLE_CPP_EXCEPTIONS": "YES",
          "GCC_ENABLE_CPP_RTTI": "YES"
	},	  
        "variables" : {
          "oci_inc_dir%" : "<!(if [ -z $OCI_INC_DIR ]; then echo \"/opt/oracle/instantclient_11_2/sdk/include/\"; else echo $OCI_INC_DIR; fi)",
          "oci_lib_dir%" : "<!(if [ -z $OCI_LIB_DIR ]; then echo \"/opt/oracle/instantclient_11_2/\"; else echo $OCI_LIB_DIR; fi)",
        }
      }
    ],
    [
      'OS=="solaris"', {
        "variables" : {
          "oci_inc_dir%" : "<!(if [ -z $OCI_INC_DIR ]; then echo \"/opt/oracle/instantclient_12_1/sdk/include/\"; else echo $OCI_INC_DIR; fi)",
          "oci_lib_dir%" : "<!(if [ -z $OCI_LIB_DIR ]; then echo \"/opt/oracle/instantclient_12_1/\"; else echo $OCI_LIB_DIR; fi)",
        }
      }
    ],
        ["OS=='win'", {
          "configurations": {
            "Release": {
              "msvs_settings": {
                "VCCLCompilerTool": {
                  "RuntimeLibrary": "2"
                }
              },
            },
            "Debug": {
              "msvs_settings": {
                "VCCLCompilerTool": {
                  "RuntimeLibrary": "3"
                }
              },
            }
          },
          "variables": {
            "oci_inc_dir%": "c:/temp/node/node_modules/oracledb/instantclient_12_1/sdk/include",
            "oci_lib_dir%": "c:/temp/node/node_modules/oracledb/instantclient_12_1/sdk/lib/msvc",
            "oci_version%": "12"
          },
          #"libraries": [ "-loci" ],
          "link_settings": {"libraries": [ '<(oci_lib_dir)\oci.lib'] }
        }]
    ],
  "cflags"        : ['-fexceptions'],
  "cflags_cc"     : ['-fexceptions'],
  "include_dirs"  : [ "<(oci_inc_dir)",
                      "src/dpi/src/",
                      "src/dpi/include/"
  ],
#  "libraries"     : ["-lclntsh"],
#  "link_settings" : {
#     "libraries"  : ['-L<(oci_lib_dir)'] 
#    }
  }
  ]
}

**Comment by cjbj** [comment]
Any chance I can get you to sign the CLA and submit a pull request? See CONTRIBUTING.md.

**Comment by rinie** [comment]
I will read the fineprint of thr CLA, although the Apache license should be
ok. Then fork, make changes in a proper way, and do the pull request
Op 24 jan. 2015 19:34 schreef "cjbj" <notifications@github.com>:

**Comment by rinie** [comment]
Mailed the OCA to oracle-ca_us@oracle.com.
Created a pull request.

Hope I did everything correct.
I had to rename node-oracledb to oracledb to get the examples working
(because of the var oracledb = require('oracledb'))

**Comment by b2dac** [comment]
Rinie great job! I tested and confirm that it works. Thank you!

**Comment by cjbj** [comment]
Fast work.  Give us a couple of days to test various windows scenarios, update doc and for your name to appear on the OCA list.

**Comment by rinie** [comment]
Ok
Op 25 jan. 2015 19:01 schreef "cjbj" <notifications@github.com>:

**Comment by iGeechee** [comment]
@rinie  Great job!  That `<(oci_lib_dir)\oci.lib'` was the piece I was missing.  I was still trying to use `oraocci<(oci_version).lib` from node-oracle while building with a full install of Oracle, but it was giving me a host of linker errors.

For anyone else, when building with a full install/local database I used the environment vars:
```
OCI_INC_DIR=C:\oracle\product\11.2.0\dbhome_1\OCI\include
OCI_LIB_DIR=C:\oracle\product\11.2.0\dbhome_1\OCI\lib\MSVC
```

I can now connect as well on Win 7 x64.

**Comment by kgajowy** [comment]
I can confirm it is working with following OS/arch:

Microsoft Windows 7 Enterprise
 6.1.7601 Service Pack 1  7601
64bit system; 32bit node/Python

**Comment by cjbj** [comment]
Bear with us a bit longer to get this merged.

@rinie is now on the OCA list, which is one hurdle done.

**Comment by zorrofox** [comment]
I use:
Microsoft Windows 7 Enterprise x64
32bit node/Python

And test for select1.js/select2.js is OK, but when I want to have some pool testing there have some issues.
```javascript
var oracledb = require('oracledb');
var dbConfig = require('./dbconfig.js');

oracledb.createPool (
  {
    user          : dbConfig.user,
    password      : dbConfig.password,
    connectString : dbConfig.connectString,
    poolMax       : 44,
    poolMin       : 2,
    poolIncrement : 5,
    poolTimeout   : 4
  },
   function(err, pool)
  {
		
		if(err) console.log(err);

		pool.getConnection (
			function(err, conn){
			  console.log(conn);
			  if(err) console.log(err);
			  conn.execute(
				  "SELECT 1 p from dual ",
				  {},
                  function(err, result){
                    if (err) console.log(err);
                    console.log(result);
                  }
			  );
		});

  });
```

Above code will exist automatically and the pool getConnection param for the anonymity function never executed. Do I miss something?

**Comment by cjbj** [comment]
@zorrofox try commenting out the line console.log(conn)

**Comment by zorrofox** [comment]
@cjbj I have tried to comment the console.log(conn), but no luck. The code just exit without any output.

**Comment by cjbj** [comment]
@zorrofox we'll look into it.

**Comment by cjbj** [comment]
The Windows pool problems is turning into a difficult to trace issue somewhere in V8/Node.js. When we resolve it we will merge the build patch.

**Comment by zorrofox** [comment]
Thank @cjbj , I just trace the function njsPool.cpp for 
```
Handle<Value> Pool::Async_AfterGetConnection(uv_work_t *req)
```
has some issue, and after this line
```
connectionBaton *connBaton = (connectionBaton*)req->data;
```
https://github.com/oracle/node-oracledb/blob/master/src/njs/src/njsPool.cpp#L511
the function has exit for exception.

**Comment by bchr02** [comment]
I just tried installing this module on another Windows PC and kept getting a long error message that ended with: node-gyp rebuild failed windows exit 1

This was on a 32 bit windows. I originally had it working on a 64 bit windows machine, so I gave it a try on another 64 bit machine. I again had the same errors.

I tried all sorts of things, to try to figure out the cause. Right before I was about to give up I tried checking what version of node I had it successfully compiled on. It was version v0.10.34 . The current version I was using was 12.

So I uninstalled node version 12 and downloaded the older version from: http://nodejs.org/dist/v0.10.34/docs/download/ and installed it.

After doing so I was able to install the module. 

Has anyone else tried installing on Windows with node version 12?

**Comment by cjbj** [comment]
We know a few calls will need to change to support 0.12.  We're looking at whether to use NAN.  See issue #2

**Comment by bchr02** [comment]
@cjbj Thanks. I didn't see that. 

Eventually as this module matures, is there any way of having it precompiled so that Visual Studios wouldn't be necessary?

**Comment by cjbj** [comment]
@bchr02 Your request is noted. 

(Any chance I can convince you to use Linux??!)

**Comment by bchr02** [comment]
I like to use Windows for developing. :grin: But for production I use both. 

Regardless, Linux or Windows, it still needs to be compiled. Although, I realize it's easier in Linux.

I will create an issue so that we can track and get community feedback.

**Comment by sagiegurari** [comment]
+1 for windows support (no need for precompiled)

**Comment by cjbj** [comment]
@zorrofox we made some headway with the pool crash.  Now we need to finalize it, review it, load test it etc.

**Comment by sagiegurari** [comment]
can you push these changes to some branch so we can see that it at least compiles fine on our systems?

**Comment by cjbj** [comment]
I will as soon as I can.

**Comment by cjbj** [comment]
I've pushed 0.3.1 - take a look.

I blogged the changes here https://blogs.oracle.com/opal/entry/node_oracledb_0_3_1

**Comment by bchr02** [comment]
great job! thank you.

**Comment by katherdx** [comment]
I skip step 6.4 because I already installed oracle in my machine.

6.5 Install the driver
If you have Instant Client in a location that is not C:\oracle\instantclient, then tell the installer where to find the Oracle libraries and headers:

(I already installed oracle in my system.)

If you are installing with a local database, use:

set OCI_LIB_DIR=C:\oracle\product\12.1.0\dbhome_1\oci\lib\msvc
set OCI_INC_DIR=C:\oracle\product\12.1.0\dbhome_1\oci\include

I didn't understand above step. how can I tell installer. where can i find ?

 

6.6 Run an example program

Set NODE_PATH to include the newly installed node-oracledb driver:

set NODE_PATH=%USERPROFILE%\AppData\Roaming\npm\node_modules


 where can i find  above path?

**Comment by bchr02** [comment]
>set OCI_LIB_DIR=C:\oracle\product\12.1.0\dbhome_1\oci\lib\msvc
>set OCI_INC_DIR=C:\oracle\product\12.1.0\dbhome_1\oci\include
>
>I didn't understand above step. how can I tell installer. where can i find ?

@katherdx  If you already have Oracle Instant Client installed then just create these two [system environmental variables](http://en.wikipedia.org/wiki/Environment_variable). [Here is a link that explains how and where to set them](http://www.itechtalk.com/thread3595.html) For this step you want to add two new system variables:
OCI_LIB_DIR needs to point to C:\oracle\product\12.1.0\dbhome_1\oci\lib\msvc
and
OCI_INC_DIR needs to point to C:\oracle\product\12.1.0\dbhome_1\oci\include

this is assuming your home directory is "C:\oracle\product\12.1.0\dbhome_1" if it's not, just adjust this path to match your location. If you are unsure what the home path is trying browsing for it. 


>6.6 Run an example program
>
>Set NODE_PATH to include the newly installed node-oracledb driver:
>
>set NODE_PATH=%USERPROFILE%\AppData\Roaming\npm\node_modules
>
>where can i find above path?

Step 6.6 - I don't understand the point of this step myself. I believe NODE_PATH is the default location where node modules are installed to when you use the ````-g```` npm install option. So it should already be set to correct path. For me, I did not install this module globally.

>cd node-oracledb
>npm install -g

So, Instead of doing the above. Navigate to the directory where you want to use the module from and do this:
>npm install c:\path\to\downloaded\module\node-oracledb

hope these steps help you.

**Comment by bchr02** [comment]
By the way, and this should be notated on the install notes, having Oracle Instant Client alone does not mean you can "Skip this step". Instead it should read "Skip this step if you already have Oracle Instant Client VERSION 12 on your machine." I say this because depending upon the version you have installed, it may not have the required vc folder. The vc folders are used by the driver. If you compile using Visual Studio 2013 for example, it will use the vc12 folder when using the module.

Another example, I have a machine with Oracle Instant Client 11.2.0.3 and it only has vc8 and vc9 folders within the msvc folder. So, using Microsoft Visual Studios versions 2013 wont work. Therefore, you should always upgrade to the latest version of Oracle Instant Client. At least version 12, which includes vc10 (for Visual Studio 2010), vc11 (2012) and vc12 (2013) folders.

I also want to mention, in case it helps someone, that I used Visual Studio 2012 for my install and it does not include Microsoft Visual C++ 2012 Redistributable and I had to install it to be able to compile.

**Comment by cjbj** [comment]
@katherdx The SET commands can be executed directly in a command shell.  This will/should automatically expand the %USERPROFILE% variable.

**Comment by bchr02** [comment]
@cjbj It should also be noted that in Windows the SET command will only set the variable for the duration of the current CMD window. Any drivers will not have access to these because they execute in a separate process. So you cannot simply use the commands: 
>set OCI_LIB_DIR=C:\oracle\product\12.1.0\dbhome_1\oci\lib\msvc
>set OCI_INC_DIR=C:\oracle\product\12.1.0\dbhome_1\oci\include

You must set them differently. [Like this](http://www.itechtalk.com/thread3595.html) or use Powershell perhaps.

**Comment by cjbj** [comment]
The OCI_INC_DIR and OCI_LIB_DIR vars are only needed during installation. If the 'npm install -g' command is run in the same cmd shell, are there any issues?

**Comment by bchr02** [comment]
Setting these two variables in this fashion did not work for me. I just though it was because there was another process spawned that didn't have access to the current CMD env variables. So, I set them in the windows system properties and it then worked.

**Comment by cjbj** [comment]
@bchr02  I'll give that feedback to the team who gave me the steps.

**Comment by bchr02** [comment]
Okay, thanks. By the way, I'm not ruling out the possibility that I did something wrong when I tried setting the variables. This is still a possibility, however I do not think this is the case. I just remember having this issue (when installing versions 0.2) and attributing it to what I new about setting variables in this fashion. Maybe I did something wrong or maybe something has changed since the last version. I don't know, let's see what your team says. Thank you.
